### DISCUSSION TITLE
Business activity monitoring (BAM) integration

### DISCUSSION THREAD

**Discussion by tokarenko** [imixs/imixs-workflow#791] [General]
Please advise how to implement business activity monitoring (BAM) with Imixs? Of particular interest is graphical visualization of process instances states and activity duration logging.

With regards, Dmitry.

**Comment by rsoika** [comment]
Hi,
Imixs-Workflow provides runtime metrics based on the [Microprofile Metric API](https://microprofile.io/project/eclipse/microprofile-metrics). This allows various ways to monitor a workflow instance. Find details [here](https://www.imixs.org/doc/engine/metrics.html).

Also take a look at [this blogpost](https://blog.imixs.org/2020/01/13/monitoring-imixs-workflow-with-grafana/) illustrating a Setup with Prometheus and Grafana. 

To get deeper into analysis of a specific workflow you can activate the Taxonomy Plugin which should be used collecting specific process-runtime information: https://www.imixs.org/doc/engine/plugins/taxonomyplugin.html . The taxonomy is based on the taxonomy features of [Lucene](https://lucene.apache.org/)

**Reply by tokarenko** [reply]
Thank you, Ralph, for your quick reply! This tool set opens up broad capabilities for analysis. The only and crucial thing I miss process instance flow visualization. I thought there was a tool available for this.

**Reply by rsoika** [reply]
What do you exactly mean by 'process instance flow visitations'? Can you provide a screen shot or picture?

**Reply by tokarenko** [reply]
Please check this example for Camunda: https://github.com/camunda-community-hub/zeebe-simple-monitor

**Reply by rsoika** [reply]
Hi,
I do not really see the benefit of such a control panel. Typically customers have thousands of process instances. It make not much sense to control them via a BPMN monitoring page. So what would be a realistic use case for such a panel?

On the other hand we are currently working on a new BPMN solution [OpenBPMN](https://github.com/imixs/open-bpmn).  This BPMN modelling tool can be integrated in various IDEs like Visual Code, Theia, Eclipse or even run in a Browser. Through the openness of the underlying [Eclipse GLSP project](https://www.eclipse.org/glsp/) it is possible to implement advanced aspects of modelling. So this may be a good approach for requirements pointing in this direction. But first I would like to see the business requirement behind this.

**Reply by tokarenko** [reply]
I think that the main use case is to provide visual aid to human operator that is better then execution logs. E.g. for a process manager to address stakeholder's questions on and issues with execution of a particular process instance; for process participant to realize what is happening, when he is going to be engaged and what is going to happen next. And so on.
You may find interesting to look at [Management and Task consoles for Kogito BPMS](https://blog.kie.org/2021/09/developing-business-processes-more-efficiently-with-runtime-tools-quarkus-extension-part-1.html) and [Video illustration](https://www.youtube.com/watch?v=SpcxTSl68RY).

I am gald to hear that you decided to embark on a new project! I wish you to achieve desired outcomes! Maybe you could consider this thread as a feature request for imixs and OpenBPMN.

**Reply by rsoika** [reply]
Hi @tokarenko, I am always a little bit surprised about this idea of how users would execute a business workflow. Below you can see a picture of a very simple "purchase order process". Such a process contains a lot of rules, decisions and executable adapters. It is impossible for a process participant to understand what is going to happen next. Via the micro-kernel architecture there can be one or more Signal Adapters involved which synchronize data, compute new information, execute business rules or even start another process.

![image](https://user-images.githubusercontent.com/1339124/178550988-c85ed6dd-e87a-43d5-af22-cf4627904a2b.png)

Even a process manager can not handle such a view if you imagine that a realistic process management system from a medium sized enterprise can have thousands of process instances to be processed either by users or by backend processes. It is the idea of a workflow management platform that processes are routed to right persons, monitored and escalated to the right manager if the situation becomes critical. It is a bad process design if a process manager or admin have to tweak the process ad hoc to solve a situation. 

What we do instead in our workflow applications, is providing a so called Workflow-History panel. This shows a human readable history of all activities, comments, subprocesses and documents in a chronological order. This is helpful for employees and also a process manager to understand what has(!) happened so far, to decide which workflow Event should be the next (e.g. approve or decline the process). 

![image](https://user-images.githubusercontent.com/1339124/178551676-ff50b6ad-1583-4697-844e-13b0bb8329db.png)

Please be aware that Imixs-Workflow is an event based human-centric workflow engine. The process is executed by employees who should provide data, approve, decline or escalate a process within the company based on compliant rules which are usually defined in a ISO 9001 based quality management. So the idea of Imixs-Wokflow is complete different from the idea of a task orientated workflow engine like camunda or activiti. 

The topic of workflow engines can be generally very complex, which is why we also offer consulting services and workshops
